### PR TITLE
feature: "addTypename" option

### PR DESCRIPTION
**Pull request by donovanhiland**
Add foundation for linting/testing/prettier/etc
Add `addTypename` option to include `__typename` in generated types
Export static version of `MockList`

### CODE REVIEW

File: src/MockFactory.ts [hunk 1]
```diff
@@ -7,18 +7,18 @@ import {
...
   addMockFunctionsToSchema,
   transformSchema,
   mergeSchemas,
 } from 'graphql-tools'
-import gql from 'graphql-tag'
 import { ResolverMap } from './types'
 import mergeResolvers from './mergeResolvers'
 
-const generateMockSchema = (schemaJson: any) => {
+const buildMockSchema = (schemaJson: IntrospectionQuery) => {
```
Comment (stevenheffner) [inline src/MockFactory.ts:21]: build def better

File: src/MockFactory.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ const generateMockSchema = (schemaJson: any) => {
     if (
       typeName.startsWith('__') ||
       typeName === 'Query' ||
+      typeName === 'Mutation' ||
```
Comment (stevenheffner) [inline src/MockFactory.ts:30]: What does this do?

File: src/MockFactory.ts [hunk 3]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
 
 class MockFactory {
   private schema: GraphQLSchema
   private mocks: ResolverMap
+  private addTypename: Boolean
 
   private mockSchema(options: { mocks?: ResolverMap } = {}) {
     const clonedSchema = transformSchema(this.schema, [])
-    const mergedMocks = mergeResolvers(this.mocks, options.mocks || {})
+    const mergedMocks = mergeResolvers(this.mocks, options.mocks ?? {})
```
Comment (stevenheffner) [inline src/MockFactory.ts:84]: Why is the `??` operator better here?

File: src/MockFactory.ts [hunk 4]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
 
-  constructor(schema: IntrospectionQuery, { mocks }: Options) {
-    this.schema = generateMockSchema(schema)
-    this.mocks = mocks || {}
+  private maybeAddTypenameToDocument = (
+    document: DocumentNode,
+    overrideAddTypename: Boolean | undefined,
+  ) => {
+    const addTypename = overrideAddTypename ?? this.addTypename
+    // ? parse(print(addTypenameToDocument(document)))
```
Comment (stevenheffner) [inline src/MockFactory.ts:94]: Left in comment ☝️

File: src/MockFactory.ts [hunk 5]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
+    schema: IntrospectionQuery,
+    { mocks, addTypename = false }: Options = {},
+  ) {
+    this.schema = buildMockSchema(schema)
+    this.mocks = mocks ?? {}
+    this.addTypename = addTypename
   }
 
-  mockFragment = <TData = any>(
+  mockFragment = <TData = any, TVariables = any>(
```
Comment (stevenheffner) [inline src/MockFactory.ts:107]: Variables for a fragment is nice.

File: src/MockFactory.ts [hunk 6]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
       throw new Error(
         'MockFactory: mockFragment only accepts fragment documents',
       )
     }
 
-    /* eslint-disable graphql/template-strings */
     const typeName = mainDefinition.typeCondition.name.value
     const fieldName = `mock__${typeName}`
-    const query = gql`
+    const query = parse(/* GraphQL */ `
```
Comment (stevenheffner) [inline src/MockFactory.ts:127]: Why this change? You no like gql tag? You like print?

File: src/MockFactory.ts [hunk 7]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
     if (!result.data || result.data[fieldName] === undefined) {
       throw new Error(
-        `Unable to generate mock data for ${
-          mainDefinition.name.value || 'fragment'
-        }. This could be a result of missing mock resolvers or an incorrect fragment structure.`,
+        [
+          `Unable to generate mock data for ${
+            mainDefinition.name.value || 'fragment'
+          }. This could be a result of missing mock resolvers or an incorrect fragment structure.`,
+          print(fragment),
```
Comment (stevenheffner) [inline src/MockFactory.ts:146]: Noice

File: src/MockFactory.ts [hunk 8]
```diff
@@ -64,111 +65,146 @@ const executeOperation = (
...
+    return JSON.parse(JSON.stringify(result.data[fieldName])) as TData
   }
 
   mockQuery = <TData = any, TVariables = any>(
     query: DocumentNode,
-    options: { variables?: TVariables; mocks?: ResolverMap } = {},
+    options: {
+      variables?: TVariables
+      mocks?: ResolverMap
+      addTypename?: true | false
```
Comment (stevenheffner) [inline src/MockFactory.ts:159]: Does the boolean type not work here?

File: src/index.ts [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import MockFactory from './MockFactory'
 
 export * from './types'
+export { MockList } from 'graphql-tools'
```
Comment (stevenheffner) [inline src/index.ts:4]: Yas queen

**Review by stevenheffner — APPROVED** [review]
I have a request. Can we make it okay to return a Mock list for a resolver

**Comment by donovanhiland** [comment]
> I have a request. Can we make it okay to return a Mock list for a resolver

@stevenheffner Care to make a PR? 😄

File: src/MockFactory.ts (hunk 6 above)
Comment (donovanhiland) [reply src/MockFactory.ts:126]: gql tag was complaining that I was creating a fragment twice. That the fragment name was already taken because I'm returning a new fragment possible with __typename on it and adding it to the bottom of this query. I don't really need `gql` and could do it without so it stopped the complaining AND removed a dependency!

File: src/MockFactory.ts (hunk 5 above)
Comment (donovanhiland) [reply src/MockFactory.ts:106]: Ya it would be except it doesn't work. Thanks for pointing this out so I can remove it for now 😢

File: src/MockFactory.ts (hunk 3 above)
Comment (donovanhiland) [reply src/MockFactory.ts:84]: I can't honestly remember why I changed this. I must have done it for a reason but I can't think of a reason the `||` wouldn't work here just as good

File: src/MockFactory.ts (hunk 2 above)
Comment (donovanhiland) [reply src/MockFactory.ts:30]: This code is responsible for adding query fields to the schema to be able to query each type. If the field that we're currently looking at is `Query` or `Mutation` then we don't do that

File: src/MockFactory.ts (hunk 3 above)
Comment (donovanhiland) [reply src/MockFactory.ts:84]: I changed it back and my memory was jogged. Eslint was complaining with the new rules I added: [prefer-nullish-coalescing](https://github.com/typescript-eslint/typescript-eslint/blob/v4.12.0/packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md)

File: src/MockFactory.ts (hunk 8 above)
Comment (donovanhiland) [reply src/MockFactory.ts:158]: Typescript was having issues when I used `Boolean` for some reason. I think I was able to work around it
